### PR TITLE
Crazy4pi314/example gallery

### LINKED ISSUE #467: Show notebooks in documentation from table of content and gallery
From #461, notebooks are included in the documentation using simple hyperlinks from the `examples.ipynb` notebook and show as html pages. However, `nbsphinx` allows to use both [table of contents](https://nbsphinx.readthedocs.io/en/0.8.0/subdir/toctree.html), RST-like links and the use of thumbnail [galleries](https://nbsphinx.readthedocs.io/en/0.8.0/subdir/gallery.html) similarly to sphinx-gallery. Implementing both would be great. The catch with respect to TOC links, is that they should be linkable also if someone is opening the `examples.ipynb` notebook, and not only looking at `examples.html` from the documentation (on Read The Docs or built locally). Probably keeping both the hyperlinks and rst-like links could work fine.

### PR DESCRIPTION
**Pull request by crazy4pi314**
Description
-----------

Same as #862, adding an example gallery to the docs. Having a # symbol in the branch name broke codecov so trying again with a re-named branch. Closes #467.

Checklist
-----------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

### CODE REVIEW

**Review by nathanshammah — COMMENTED** [review]
Does the `Docker compose` necessarily need to be added before adding the gallery? I would prune it and open a specific PR only for that infrastructure modification.

File: docs/source/conf.py [hunk 1]
```diff
@@ -322,3 +324,21 @@ def get_incollection_template(self, e):
...
+nbsphinx_thumbnails = {
+    'examples/ibmq-backends': '_images/unitary_fund_logo.png',
+    'examples/hamiltonians': '_images/vqe-cirq-pauli-sum-mitigation-plot.png',
+    'examples/simple_landscape': '_images/simple_landscape_13_0.png',
+    'examples/maxcut-demo': '_images/maxcut-demo_48_1.png',
+    'examples/pyquil_demo': '_images/unitary_fund_logo.png',
+    'examples/cdr_api': '_images/unitary_fund_logo.png',
+    'examples/mitiq-paper/mitiq-paper-codeblocks':
+        '_images/unitary_fund_logo.png',
+}
```
Comment (rmlarose) [inline docs/source/conf.py:344]: I think you need to add these images?

Q: What happens if we add a new example and don't add an entry to this dict? Does it default to some standard image? Is it possible to default to `unitary_fund_logo.png`?

**Comment by crazy4pi314** [comment]
OK, should be good to go now!

File: docs/source/examples/examples.myst [hunk 2]
```diff
@@ -1,15 +1,11 @@
...
 
-```{toctree}
----
-maxdepth: 2
----
-ibmq-backends.rst
-hamiltonians.rst
-simple_landscape.rst
+```{nbgallery}
+Defining an Mitiq executor with a hamiltonian <hamiltonians.rst>
```
Comment (andreamari) [inline docs/source/examples/examples.myst:4]:
```suggestion
Defining a Mitiq executor with a hamiltonian <hamiltonians.rst>
```

File: docs/source/examples/examples.myst [hunk 3]
```diff
@@ -1,15 +1,11 @@
...
+Variational circuit mitigation with zero noise extrapolation  <simple_landscape.rst>
 maxcut-demo.myst
-pyquil_demo.myst
+Understanding probabilistic error cancellation (PEC) <pec-tutorial.myst>
+Zero noise extrapolation on IBMQ backends<ibmq-backends.rst>
+Zero noise extrapolation on pyQuil parametric programs <pyquil_demo.myst>
```
Comment (andreamari) [inline docs/source/examples/examples.myst:9]:
```suggestion
Zero noise extrapolation on PyQuil parametric programs <pyquil_demo.myst>
```

**Review by andreamari — APPROVED** [review]
Small corrections. LGTM!

File: docs/source/conf.py (hunk 1 above)
Comment (rmlarose) [reply docs/source/conf.py]: This is better, thanks!

**Review by rmlarose — APPROVED** [review]

**Comment by crazy4pi314** [comment]
Thanks all!

**Pull request merged**